### PR TITLE
Add .toBuffer() to return a node.js Buffer

### PR DESCRIPTION
**Pull request by axic**
As the constructor already supports Buffer as an input, it would make sense to make this bidirectional.

Buffer+BN together is used extensively in ethereumjs-vm and ethereumjs-abi to encode/decode bytecode.

Added the .toBuffer() as a new feature which asserts on Buffer being present. Comes with its own test.

Also added tests for using Buffer as an input to the constructor.

### CODE REVIEW

**Comment by jprichardson** [comment]
:+1:

File: lib/bn.js [hunk 1]
```diff
@@ -52,6 +52,7 @@ else
 
 BN.BN = BN;
 BN.wordSize = 26;
+BN._hasBuffer = typeof Buffer !== 'undefined';
```
Comment (indutny) [inline lib/bn.js]: How does it play with browserify? Will it include `Buffer` for browsers?

Comment (axic) [inline lib/bn.js]: Good catch, just checked and it does include the browserify version of Buffer.

However ignoring that module alleviates it: ```browserify --ignore buffer lib/bn.js -s BN -o bn-browser.js```

Comment (indutny) [inline lib/bn.js]: I'm not sure if it will work well when `bn.js` is used as dependency. Usually, I work-around such problems by doing `require('buf' + 'fer').Buffer` inside of `try { } catch {}`.

In this case, though, such check should happen lazily, IMO.

File: lib/bn.js [hunk 2]
```diff
@@ -459,6 +460,11 @@ BN.prototype.toJSON = function toJSON() {
   return this.toString(16);
 };
 
+BN.prototype.toBuffer = function toBuffer(endian, length) {
```
Comment (axic) [inline lib/bn.js]: Sounds good, but Buffer is not an exact match. ```push``` and ```unshift``` are not available.

Comment (indutny) [inline lib/bn.js]: Yeah, it will require a bit of rewriting of code below, but I think it should be possible. Are you interested in working on this?

Comment (axic) [inline lib/bn.js]: I am not entirely sold that ```toArrayLike(Buffer, ...)``` would be handier to use than ```new Buffer(bn.toArray(...))```.

However BN itself could export an index method and therefore become indexable as an Array?

Or alternatively it could make sense to have a node-bn.js package which has these kind of extensions? Although toBuffer on its own might be too slim for that.

Comment (indutny) [inline lib/bn.js]: I didn't propose to make it an external API. Rather a common part for both `toArray` and `toBuffer`.

Comment (indutny) [inline lib/bn.js]: `toBuffer` seems to be very minor.

Comment (axic) [inline lib/bn.js]: Ah ok, I thought it is part of the dependency problem for browserify, because using it like the above solves that issue.

File: lib/bn.js (hunk 1 above)
Comment (axic) [inline lib/bn.js]: Pushed an implementation based on the suggested. It seems to work with browserify, the tests and an actual web browser.

File: lib/bn.js (hunk 2 above)
Comment (axic) [inline lib/bn.js]: I think implementing ```toArrayLike``` should be left as an exercise for a separate pull request, because it is about optimisation.

File: lib/bn.js [hunk 3]
```diff
@@ -459,6 +465,11 @@ BN.prototype.toJSON = function toJSON() {
...
+  assert(typeof BN._Buffer !== 'undefined');
```
Comment (indutny) [inline lib/bn.js]: Is there any point in exposing `_Buffer` to users? What do you think about making at a context variable instead?

File: lib/bn.js [hunk 4]
```diff
@@ -53,6 +53,12 @@ else
 BN.BN = BN;
 BN.wordSize = 26;
 
+try {
```
Comment (indutny) [inline lib/bn.js]: Like declaring variable here.

File: lib/bn.js (hunk 2 above)
Comment (indutny) [inline lib/bn.js]: Sorry, but I disagree. Most of the functions here were created with the aim for performance, and introducing `toBuffer` means certain expectations from the API users. I would appreciate if we could meet this expectations and make it as fast as we can.

Thanks!

File: lib/bn.js (hunk 3 above)
Comment (axic) [inline lib/bn.js]: What should be the variable name? Ideas:
- _Buffer or __Buffer
- nodeBuffer
- nodejsBuffer
- BufferCons
- BufferAPI

I think I would go with nodeBuffer or nodejsBuffer.

Comment (indutny) [inline lib/bn.js]: just `Buffer`, IMO.

File: lib/bn.js [hunk 5]
```diff
@@ -53,6 +53,12 @@ else
...
+  BN._Buffer = require('buf' + 'fer').Buffer;
```
Comment (odeke-em) [inline lib/bn.js]: Just out of curiosity why
```javascript
require('buf' + 'fer')
```
instead of
```javascript
require('buffer');
```

Comment (indutny) [inline lib/bn.js]: Because of `browserify`.

Comment (odeke-em) [inline lib/bn.js]: Oh wow.

Comment (Kagami) [inline lib/bn.js]: Like I already mentioned in [PR to your other project](https://github.com/indutny/brorand/pull/3), webpack is (unfortunately) smart enough to break this hack.

**Comment by dcousens** [comment]
Out of interest,  @indutny what are you trying to prevent?

File: lib/bn.js (hunk 5 above)
Comment (dcousens) [inline lib/bn.js]: Out of interest,  @indutny what are you trying to prevent?

Comment (indutny) [inline lib/bn.js]: I don't want to bundle `buffer` when linking this module in browserify.

File: lib/bn.js (hunk 2 above)
Comment (axic) [inline lib/bn.js]: I still believe this should be part of a separate pull request as it involves more changes.

The first step is https://github.com/indutny/bn.js/pull/81 to remove the push/unshift dependency. After that it is simple.

Comment (axic) [inline lib/bn.js]: Based on the merged #81, the following patch works fine:

```diff
diff --git a/lib/bn.js b/lib/bn.js
index b657b34..7787c04 100644
--- a/lib/bn.js
+++ b/lib/bn.js
@@ -467,17 +467,21 @@ BN.prototype.toJSON = function toJSON() {
 
 BN.prototype.toBuffer = function toBuffer(endian, length) {
   assert(typeof Buffer !== 'undefined');
-  return new Buffer(this.toArray(endian, length));
+  return this.toArrayLike(Buffer, endian, length);
 };
 
 BN.prototype.toArray = function toArray(endian, length) {
+  return this.toArrayLike(Array, endian, length);
+};
+
+BN.prototype.toArrayLike = function toArray(cons, endian, length) {
   var byteLength = this.byteLength();
   var reqLength = length || byteLength;
   assert(byteLength <= reqLength, 'byte array longer than desired length');
 
   this.strip();
   var littleEndian = endian === 'le';
-  var res = new Array(reqLength);
+  var res = new cons(reqLength);
 
   var q = this.clone();
   if (!littleEndian) {
```

**Comment by axic** [comment]
@indutny do you think this can be merged now?

File: lib/bn.js (hunk 5 above)
Comment (axic) [inline lib/bn.js]: @indutny @Kagami I've came across this version of instructing browser of non inclusion (from here: https://github.com/jprichardson/secure-random/commit/bccdbd2004c0895f8a3a382db19f5aac7643a6a1), might worth a try?

It works by adding a section to `package.json`

**Comment by indutny** [comment]
Hell yes, this looks great!

**Comment by axic** [comment]
@indutny I did a rebase against master, should I force push it to this PR (or should I create a new PR) so you can merge it easily?

File: lib/bn.js (hunk 2 above)
Comment (axic) [inline lib/bn.js]: Since `semistandard` is used now, the above will fail with: `bn.js:510:19: A constructor name should not start with a lowercase letter.` The parameter must be named `Cons` to make the `new` keyword happy.

**Comment by indutny** [comment]
Force push should be fine, assuming that you didn't do merge commits in your branch.

**Comment by axic** [comment]
Done, there should be no conflicts

**Comment by indutny** [comment]
Hm... I thought that there was suggestion about using `browser` field in package.json. What do you think about this @axic ?

**Comment by axic** [comment]
@indutny I suggested it, but there was no response

I tested it and works, at least the browserified output is the same size, but not an exact copy.

```js
},{"buffer":2}],2:[function(require,module,exports){
```

It includes an empty package named `buffer`. With the try/catch it works.

**Comment by indutny** [comment]
Great, let's do it this way. Sorry for not responding!

**Comment by axic** [comment]
Pushed the package.json way

**Comment by indutny** [comment]
PR no longer applies... May I ask you to rebase?

**Comment by axic** [comment]
@indutny github said there are no conflicts, but rebased anyway

**Comment by indutny** [comment]
Landed, thank you! (Released too)
